Post-check algebraic and coupling ghosting enhancements

lindsayad opened the pull request:
This marks the importance of local testing. 1 layer of sparsity was not actually getting added in a Rattlesnake simulation that added all its dg/interface kernels through actions. The issue was that the logic in Action::addRelationshipManager had determined that because the added RM had an RM type of GEOMETRIC | ALGEBRAIC | COUPLING that the RM must have been added during the geometric attachment stages and hence did not add it for the requested `input_rm_type = COUPLING`. Now in our `CouplingFunctorCheckAction` I make sure that we add an RM with rm type alternately restricted only to algebraic and then coupling.

Note that this logic presumes that the user has added a relationship manager that provides geometric ghosting ahead of this coupling ghosting. If not, the user will be in for a rude awakening if they try to run with DistributedMesh. Note that we can't try to fix this situation for them; by the time we've added all our compute objects to the warehouse, the mesh has been prepared and remote elements have already been deleted.

Initially I only added a coupling functor and then reinitialized the matrix sparsity, but one quick test in parallel revealed that that was quite stupid... An object that needs a layer of sparsity implies that it also needs a layer of algebraic ghosting. This we can also set-up for the user intelligently if they did not provide their own algebraic ghosting. Sadly for the user, this will require some potentially expensive calls localize the solution vectors, re-build the send lists, and then parallelize again. Motivation to them to add the correct relationship managers in their actions!

lindsayad committed:
Make sure that 1 layer coupling is actually added

This marks the importance of local testing. 1 layer
of sparsity was not actually getting added in a Rattlesnake
simulation that added all its dg/interface kernels through
actions. The issue was that the logic in Action::addRelationshipManager
had determined that because the added RM had an RM type of
GEOMETRIC | ALGEBRAIC | COUPLING that the RM must have been
added during the geometric attachment stages and hence did not
add it for the requested `input_rm_type = COUPLING`. Now in
our `CouplingFunctorCheckAction` I make sure that we add an
RM with rm type restricted only to coupling, as the name of
the action implies. Note that this logic presumes that the user
has added a relationship manager that provides geometric
ghosting ahead of this coupling ghosting. If not, the user will
be in for a rude awakening if they try to run with DistributedMesh.
Note that we can't try to fix this situation for them; by the time
we've added all our compute objects to the warehouse, the mesh
has been prepared and remote elements have already been deleted.

Refs #13736

lindsayad committed:
Also add 1 layer algebraic ghosting for dgs and iks

It would be useless to add 1 level of sparsity without also
adding 1 level of ghosting to the solution vector because
to otherwise we get access into vector elements that don't
exist on the process.

lindsayad commented:
@YaqiWang

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/375264/) on 6069b64 : invalidated by @lindsayad

lindsayad committed:
Add no functor addition test

Demonstrates that we do not add any additional rms
if a MooseObjectAction has already added an rm with
sufficient ghosting coverage

lindsayad committed:
Show geometric ghosting failure

This requires first actually turning off default mesh
ghosting in MooseMesh. I will need to reconfigure this
to somehow be user settable as default_ghosting is a param
in FEProblemBase. The issue here is that the FEProblemBase
object is created after the mesh has been prepared and remote
elements have been deleted.

Anyways, when we remove default mesh ghosting, we show that
when some object requires algebraic ghosting the simulation will
fail because algebraic ghosting requires corresponding geometric
ghosting. And unfortunately the CouplingFunctorCheckAction just
runs too late to provide geometric ghosting. This means
the responsibility falls on the user

lindsayad commented on framework/src/mesh/MooseMesh.C:
```diff
@@ -2048,6 +2048,7 @@ MooseMesh::buildMeshBaseObject(ParallelType override_type)
 
     mesh = libmesh_make_unique<ReplicatedMesh>(_communicator, dim);
   }
+  mesh->remove_ghosting_functor(mesh->default_ghosting());
```
@permcody @friedmud I guess that #13145 really didn't remove default ghosting haha. Looks like there's only [24 tests failing](https://civet.inl.gov/job/375328/) though, so we should be able to truly remove default ghosting pretty soon!

lindsayad committed:
Revert "Show geometric ghosting failure"

This reverts commit 5a36ebe03c00b42d57378565d5f4eb7cb79559da.

YaqiWang commented on test/src/actions/AddDGDiffusion.C:
```diff
@@ -0,0 +1,65 @@
+  MooseEnum kernels_to_add("BOTH REGULAR COUPLED", "BOTH");
+  params.addParam<MooseEnum>(
+      "kernels_to_add", kernels_to_add, "What dgkernels to add from this action");
+
```
So this action does not need to worry about adding relationship managers even it is adding DG kernels?

lindsayad replied:
It should, but what we are testing here is whether the `CouplingFunctorCheckAction` can save developers who **did not** properly add ghosting in their actions. The `CouplingFunctorCheckAction` is capable of saving developers who forgot algebraic or coupling ghosting, but it cannot save them if they/we forgot geometric ghosting

lindsayad replied:
So it's worth discussing (@permcody @friedmud ) whether we want to keep default geometric ghosting, which has been happening behind the scenes even after #13145

YaqiWang replied:
I do not think geometric ghosting is costly. Only because of using PETSc, we do not it in some circumstances if I understand correctly. In general, we should keep geometric ghosting in my opinion. As a developer, it will be nice if I do not need to worry about algebraic and coupling ghosting.

permcody replied:
I think we should always minimize ghosting. No need to keep something that isn't used. @YaqiWang - what about a a relatively simple equation system with adaptivity turned on? Adaptivity can be very expensive and with ghosting on, you increase communication at every step.

lindsayad replied:
Adaptivity requires ghosting though...

YaqiWang replied:
I do not think geometric ghosting requires any communication during the solve. It is about how we handle the mesh distribution. It is done where mesh is setup for use.

permcody replied:
and then with each adaptive step...

YaqiWang replied:
Right. But geometric ghosting costs much less than others both in CPU time and memory. It is not worthwhile to spend much effort to optimizing it probably.

permcody replied:
I would agree with you if it were a question of optimization. It's more a question of correctness. If it's as simple as turning something off (which we thought was off anyway), and things still work - I don't see why not.

lindsayad replied:
It would be good to conjure up a problem where geometric ghosting is unnecessary and where we think it might be expensive. I'm struggling to think of something that would satisfy the latter condition. In a regular simulation case the mesh should only run geometric ghosting functors and delete remote elements once, when `prepare_for_use` is called. Adaptivity would run the functors every time step, but then those functors are actually needed.

The problem with ***not*** having default geometric ghosting really only arises for a small sub-set of MOOSE users...those users who generate all their `MooseObjects` through custom actions, thus avoiding the `MooseObjectAction::addRelationsipManagers` call. For this small sub-set of users, they must either have default geometric ghosting or they must override `addRelationshipManagers(input_rm_type)` in their custom action.

YaqiWang replied:
If this is a problem, how about let application call something to indicate having the default geometric ghosting? For example, we can call something in RattlesnakeApp.

lindsayad replied:
I'm on board for something like that

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/375455/) on fda2550 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13838/site)

This comment will be updated on new commits.

permcody commented on framework/src/actions/Action.C:
```diff
@@ -195,28 +201,29 @@ Action::addRelationshipManagers(Moose::RelationshipManagerType input_rm_type,
+    // ensuring that a non-null CouplingMatrix (reflecting coupling in the nonlinear system) doesn't
+    // get applied to the auxiliary system
+
+    bool is_algebraic = (rm_type & Moose::RelationshipManagerType::ALGEBRAIC) ==
```
We have the shortcut methods in RelationshipManager.h (`isType`, `getType`, etc)

lindsayad replied:
Those require having an instance which I do not yet have here

permcody replied:
static methods?

lindsayad replied:
That takes two arguments to distinguish it from the existing single argument case? I guess I can do that although I'll have to figure out a good method name to clearly distinguish which argument will used in the `&` operation and as the comparator.

lindsayad replied:
Ah I got ya...yea maybe a collection of static methods like `isAlgebraic(rm_type)`, `isGeometric(rm_type)`...I like that

permcody commented on framework/src/actions/Action.C:
```diff
@@ -195,28 +201,29 @@ Action::addRelationshipManagers(Moose::RelationshipManagerType input_rm_type,
+                        Moose::RelationshipManagerType::ALGEBRAIC;
+    bool is_coupleable = (rm_type & Moose::RelationshipManagerType::COUPLING) ==
+                         Moose::RelationshipManagerType::COUPLING;
+    auto sys_type = (is_algebraic || is_coupleable) ? Moose::RMSystemType::NONLINEAR
```
Don't necessarily have all of these, they should probably be migrated to RelationshipManager so they can be in one central location.

lindsayad replied:
Are you talking about class methods? Don't have an instance yet here

permcody replied:
Yes

permcody commented on test/src/actions/AddDGDiffusion.C:
```diff
@@ -0,0 +1,65 @@
+    params.set<Real>("epsilon") = _pars.get<Real>("epsilon");
+    if (_pars.isParamValid("diff"))
+      params.set<MaterialPropertyName>("diff") = _pars.get<MaterialPropertyName>("diff");
+    _problem->addDGKernel("DGDiffusion", "dg_diffusion", params);
```
Use `InputParameters::apply*` methods here

permcody commented:
Can you take a look at the `DisplayGhostingAction` object and make sure it works properly with your CouplingGhosting stuff.

lindsayad committed:
Expand relationship manager ability and nuance

This commit does several things:
- Makes action inherit from ParallelObject so we have
  a communicator
  - With this in hand, we can choose not to add rms if
    we are not actually running a parallel simulation
- Expand the RelationshipManager interface by adding
  several static methods and variables that make it easy
  to quickly create rm types and to check whether rm types
  are geometric, etc.

lindsayad committed:
Explicitly test what we are reinitializing

Parsing the console output, check whether we are
reinitializing vectors and sparsity when it is needed
only

lindsayad committed:
Move MaxVarNDofsPerElem calculation

Previously this was in NonlinearSystemBase::init
which gets called before our computing objects get added.
The problem with this location is that it didn't allow
us to run the MaxVarNDofsPerElem calculation intelligently.
That object exists only for AD purposes, hence if we can
move its calculation to NonlinearSystemBase::intialSetup
then we can see whether we actually have any AD objects. If
we don't then we don't need to run the MaxVarNDofsPerElem
calculation. This was showing up as the primary computational
expense in one rattlesnake simulation, where it served no purpose.

Tagging @YaqiWang because he should appreciate this speed-up

lindsayad committed:
Try sneaking in a mortar error check

lindsayad commented:
I don't understand this minimum gcc version failure. I cannot reproduce with min-moose on rod.

moosebuild commented:
Job [Minimum gcc version](https://civet.inl.gov/job/376980/) on a080981 : invalidated by @lindsayad

lindsayad commented:
Alright cool, the previously failing test passed but the other test failed this time 😭 There obviously has to be something going on with the timed print...

lindsayad commented:
BTW, @permcody @friedmud `TimedPrint` is some really cool code. Just that class could be used as a one hour demo for advanced c++ features if we're ever doing our c++ tutorial.

permcody commented:
Wait did you figure it out? Kind of conflicting messages here

On Fri, Aug 2, 2019 at 12:20 PM Alex Lindsay <notifications@github.com>
wrote:

> BTW, @permcody <https://github.com/permcody> @friedmud
> <https://github.com/friedmud> TimedPrint is some really cool code. Just
> that class could be used as a one hour demo for advanced c++ features if
> we're ever doing our c++ tutorial.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13838?email_source=notifications&email_token=AAXFOICI3A3GDSZVGXX3ZOLQCR3G3A5CNFSM4IICAMVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3OPZBQ#issuecomment-517799046>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXFOIHMBHP53LBMSI5JKILQCR3G3ANCNFSM4IICAMVA>
> .
>

lindsayad committed:
Try to always print for duration zero

lindsayad commented:
> Wait did you figure it out? Kind of conflicting messages here

The issue is that I want to get the behavior of `CONSOLE_TIMED_PRINT`, but I want the message to print "instantly" or no matter how fast it takes to run the section, e.g. if it takes a milisecond to reinitialize the vectors or the sparsity I want my message printed that we're reinitializing because it is an indicator to developers/users whether they got their ghosting correct or whether our post-checks had to do some work for them.

To achieve this behavior, I am passing an `initial_wait` of `std::chrono::duration<double>(0)`. This appears to work in all cases except for the minimum gcc target. Apparently on that target the reinitialization happens so fast that the result of `done_future.wait_for(initial_wait)` is not a `std::future_status::timeout` which seems insane because the `initial_wait` is zero. However, perhaps the start-up of the thread itself takes some non-zero amount of time...?

lindsayad commented:
So @YaqiWang's stuff that uses `AMGUpdate` only started malloc'ing after #13750 because he attaches the `DofMap` from the nonlinear system which previously to #13750 had a default coupling functor attached to it. However, anything that uses `PhysicsBasedPreconditioner` will have been malloc'ing ever since #13145 because for PBP we create brand new systems with brand new DofMap's and since we disabled default ghosting in the `EquationSystems` object these new systems DofMap's don't have any default coupling functor attached to them.

lindsayad commented:
My vote for design is this:

- Do not disable ghosting using `EquationSystems::enable_default_ghosting`
- Instead we explicitly disable only ***algebraic ghosting*** for all libMesh Systems created within the MOOSE repository
- We ***keep*** the default coupling functor in all Systems. This means `PBP` and @YaqiWang's `AMGUpdate` will be good to go out of the box. Similarly for future application developers who do some out of the ordinary things
- We have the `CouplingFunctorCheckAction` that is capable of rescuing application developers who add all their `MooseObjects` through actions ***and*** use objects that we know need an additional layer of algebraic ghosting/coupling. This will be at the expense of vector/sparsity `reinit` but it's better than a bunch of mallocs.

lindsayad committed:
Don't remove default ghosting through EquationSystems

Instead manually remove it through the DofMap and through
the MeshBase objects. Do not remove the zero-level default
coupling functor.

permcody commented on framework/include/utils/TimedPrint.h:
```diff
@@ -26,15 +26,17 @@
 #endif
 
-#define CONSOLE_TIMED_PRINT(...)                                                                   \
+#define CONTROLLED_CONSOLE_TIMED_PRINT(initial_wait, dot_interval, ...)                            \
```
Is this necessary? I think you can already pass in the initial and wait periods to the existing macro.

permcody commented on framework/src/relationshipmanagers/RelationshipManager.C:
```diff
@@ -79,7 +79,63 @@ RelationshipManager::RelationshipManager(const InputParameters & parameters)
+    Moose::RelationshipManagerType::COUPLING;
+
+InputParameters
+dummyParams()
```
empty and dummy? Lol

lindsayad commented:
@YaqiWang This should be in your rattlesnake moose submodule now. How are things doing?

YaqiWang commented:
`test:actions/saaf_sn_cfem.array_action_vacuum1_test` test in debug mode takes 273s on equation system initialization, should we be concerned? Good news is that the tutorial one runs fine now.

YaqiWang commented:
Sorry for being a little late, I was in a conversation.

lindsayad commented:
Can you give me a civet link where I can see that slow test? I just ran it
in oprof mode on my local machine and it ran in 1.06 seconds, and building
the sparsity pattern was only the 5th heaviest function in the resulting
profile.

On Tue, Aug 6, 2019 at 10:38 AM Yaqi <notifications@github.com> wrote:

> Sorry for being a little late, I was in a conversation.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13838?email_source=notifications&email_token=ACOGA4EDVYYWQNB5X6L7NPDQDGZJ3A5CNFSM4IICAMVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3V45PQ#issuecomment-518770366>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACOGA4BTDFR7H4CBDMYDU53QDGZJ3ANCNFSM4IICAMVA>
> .
>

YaqiWang commented:
https://moosebuild.inl.gov/job/197828/

lindsayad commented:
It's interesting how slow it is on the test box. It took 12 seconds in
debug mode on my laptop with the exact same rattlesnake, moose, and libmesh
commits.

I guess I would not be concerned. I ran the profiler again with those
matching commits. 56% of the run time is spent in the solve. 9.1% is in the
EquationSystems::init, with 89% of that time in the
SparsityPattern::handle_vi_vj method. I don't see anything odd going on in
the sparsity computation method...the correct level of elements is being
ghosted: 0, meaning just intra-element dof coupling. And the coupling
matrix is non-null so we are applying our coupling matrix from MOOSE. The
SparsityPattern::handle_vi_vj method gets called for every variable and you
have 40, so I feel like the timing is reasonable.

With the change back to use libMesh's default coupling, we get this
sparsity pattern calculation done in EquationSystems::init. If we had not
moved back to this, we would still have the same computational expense, but
it would occur during an ImplicitSystem::reinit() called from
CouplingFunctorCheckAction::act().

This is all to say that I don't think this particular test should be any
slower than if it was run prior to #13145. I could try to go back that far
and run a profile from pre-13145 if you want me to.

On Tue, Aug 6, 2019 at 1:09 PM Yaqi <notifications@github.com> wrote:

> https://moosebuild.inl.gov/job/197828/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13838?email_source=notifications&email_token=ACOGA4BA2ODEA3JJRWUDAYLQDHLBDA5CNFSM4IICAMVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3WKFCI#issuecomment-518824585>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACOGA4HCIIATFEQQWSUYGSLQDHLBDANCNFSM4IICAMVA>
> .
>

YaqiWang commented:
That is odd. Interestingly retesting of that timeout test passed (check out that link) that I did not notice before. It could be a random timeout failure or the runtime is close to the limit (300s). I have no idea how on the test box the debug run can that slower. If it is simple for you to have the same setup as the test box, I want a double-check. Otherwise it is fine to me. I will add some more benchmark tests in Rattlesnake and make sure we will get an test failure when the run-time increased by 5% for any new PR.

permcody commented:
Are we comparing the laptop (Mac) versus the build box (Linux)? If so, we
have seen this many times before. Linux is much slower with the extra
container bounds checking and other Linux-only checks that run versus Mac.
Still Thanks Alex for checking into this.

The build box timeout retesting is indeed interesting. I look at them
routinely and am often surprised about how much difference the same test
can take when run with the group and re-run in the timeout scenario. It can
be more than ~30s or 10% difference in runtime. My only guess as to what
might be happening is that those build boxes are terrible at handling the
multiple jobs we throw at them (-j 8 or 16). When we rerun the timeouts,
it's usually just one or two tests that run by themselves.

On Tue, Aug 6, 2019 at 8:05 PM Yaqi <notifications@github.com> wrote:

> That is odd. Interestingly retesting of that timeout test passed (check
> out that link) that I did not notice before. It could be a random timeout
> failure or the runtime is close to the limit (300s). I have no idea how on
> the test box the debug run can that slower. If it is simple for you to have
> the same setup as the test box, I want a double-check. Otherwise it is fine
> to me. I will add some more benchmark tests in Rattlesnake and make sure we
> will get an test failure when the run-time increased by 5% for any new PR.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13838?email_source=notifications&email_token=AAXFOICRJJ53YBQINQDH6ETQDIGVJA5CNFSM4IICAMVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD3WZTBY#issuecomment-518887815>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXFOIH2YQTS5D3X2Q57Q73QDIGVJANCNFSM4IICAMVA>
> .
>

lindsayad commented:
> Are we comparing the laptop (Mac) versus the build box (Linux)?

Yes

> If so, we have seen this many times before. Linux is much slower with the extra container bounds checking and other Linux-only checks that run versus Mac. Still Thanks Alex for checking into this.

Aw, I appreciate you pointing out the differences. That at least makes some sense to me now.

lindsayad commented:
The test takes 2 minutes and 38 seconds in debug mode on cone

lindsayad commented:
I also ran the test with the first rattlesnake merge containing your test (with corresponding moose and libmesh commits) and the test took 2 minutes and 36 seconds. So this test was not slowed down by this PR; it has always been on the edge.

friedmud commented:
They should add relationship managers for their actions if they need
ghosting.  But that's just my vote.

On Wed, Jul 31, 2019 at 5:50 PM Alex Lindsay <notifications@github.com>
wrote:

> *@lindsayad* commented on this pull request.
> ------------------------------
>
> In test/src/actions/AddDGDiffusion.C
> <https://github.com/idaholab/moose/pull/13838#discussion_r309476174>:
>
> > +validParams<AddDGDiffusion>()
> +{
> +  auto params = validParams<Action>();
> +  params.addRequiredParam<NonlinearVariableName>("variable",
> +                                                 "The variable on which the dgkernel will act");
> +  params.addParam<Real>("sigma", "sigma");
> +  params.addParam<Real>("epsilon", "epsilon");
> +  params.addParam<MaterialPropertyName>(
> +      "diff", "The diffusion (or thermal conductivity or viscosity) coefficient.");
> +  params.addParam<VariableName>("coupled_var",
> +                                "The coupled variable that will determine the diffusion "
> +                                "rate from the DGCoupledConvection dgkernel");
> +  MooseEnum kernels_to_add("BOTH REGULAR COUPLED", "BOTH");
> +  params.addParam<MooseEnum>(
> +      "kernels_to_add", kernels_to_add, "What dgkernels to add from this action");
> +
>
> It would be good to conjure up a problem where geometric ghosting is
> unnecessary and where we think it might be expensive. I'm struggling to
> think of something that would satisfy the latter condition. In a regular
> simulation case the mesh should only run geometric ghosting functors and
> delete remote elements once, when prepare_for_use is called. Adaptivity
> would run the functors every time step, but then those functors are
> actually needed.
>
> The problem with *not* having default geometric ghosting really only
> arises for a small sub-set of MOOSE users...those users who generate all
> their MooseObjects through custom actions, thus avoiding the
> MooseObjectAction::addRelationsipManagers call. For this small sub-set of
> users, they must either have default geometric ghosting or they must
> override addRelationshipManagers(input_rm_type) in their custom action.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/13838?email_source=notifications&email_token=AAGUSMK7OKTL2YCJUVMLTWDQCIQNDA5CNFSM4IICAMVKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCAG4LLI#discussion_r309476174>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGUSMI6DODVK2CTZARTTNTQCIQNDANCNFSM4IICAMVA>
> .
>
